endpoints/openrtb2/amp_auction.go

```diff
@@ -0,0 +1,236 @@
+				// Looking for cache_id to be set, as this should only be set on winning bids (or
+				// deal bids), and AMP can only deliver cached ads in any case.
+				// Note, this could casue issues if a targeting key value starts with "hb_cache_id",
+				// but this is a very unlikely corner case. Doing this so we can catch "hb_cache_id"
+				// and "hb_cache_id_{deal}", which allows for deal support in AMP.
+				bidExt := &openrtb_ext.ExtBid{}
+				err := json.Unmarshal(bid.Ext, bidExt)
+				if err != nil {
+					w.WriteHeader(http.StatusInternalServerError)
+					fmt.Fprintf(w, "Critical error while unpacking AMP targets: %v", err)
+					glog.Errorf("/openrtb2/amp Critical error unpacking targets: %v", err)
+					return
+				}
+				for key, value := range bidExt.Prebid.Targeting {
+					targets[key] = value
+				}
+			}
+		}
+	}
+
+	// Now JSONify the tragets for the AMP response.
+	ampResponse := AmpResponse{
+		Targeting: targets,
+	}
+
+	// Fixes #231
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(false)
+
+	// If an error happens when encoding the response, there isn't much we can do.
+	// If we've sent _any_ bytes, then Go would have sent the 200 status code first.
+	// That status code can't be un-sent... so the best we can do is log the error.
+	if err := enc.Encode(ampResponse); err != nil {
+		glog.Errorf("/openrtb2/amp Error encoding response: %v", err)
+	}
+}
+
+// parseRequest turns the HTTP request into an OpenRTB request.
+// If the errors list is empty, then the returned request will be valid according to the OpenRTB 2.5 spec.
+// In case of "strong recommendations" in the spec, it tends to be restrictive. If a better workaround is
+// possible, it will return errors with messages that suggest improvements.
+//
+// If the errors list has at least one element, then no guarantees are made about the returned request.
+func (deps *endpointDeps) parseAmpRequest(httpRequest *http.Request) (req *openrtb.BidRequest, errs []error) {
+	// Load the stored request for the AMP ID.
+	req, errs = deps.loadRequestJSONForAmp(httpRequest)
+	if len(errs) > 0 {
+		return
+	}
+
+	// Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
+	deps.setFieldsImplicitly(httpRequest, req)
+
+	// Need to ensure cache and targeting are turned on
+	errs = enforceAMPCache(req)
+	if len(errs) > 0 {
+		return
+	}
+
+	// At this point, we should have a valid request that definitely has Targeting and Cache turned on
+
+	if err := deps.validateRequest(req); err != nil {
+		errs = []error{err}
+		return
+	}
+	return
+}
+
+// Load the stored OpenRTB request for an incoming AMP request, or return the errors found.
+func (deps *endpointDeps) loadRequestJSONForAmp(httpRequest *http.Request) (req *openrtb.BidRequest, errs []error) {
+	req = &openrtb.BidRequest{}
+	errs = nil
+
+	ampId := httpRequest.FormValue("tag_id")
+	if len(ampId) == 0 {
+		errs = []error{errors.New("AMP requests require an AMP config ID")}
```

Comment:
Do you think it'll be clear to callers that "AMP config ID" is the same as the "tag_id" query param? I'm not so sure...

I know that config_id => tag_id was a last-minute rename... might be worth clarifying some of these error messages in light of it though.

Reply:
Will fix all the error messages.